File: imu_calib_python/imu_calib/runtime/print_calibration_summary.py
from __future__ import annotations

import numpy as np

from imu_calib.runtime.result_summary import get_allan_status, get_temperature_model_status


def print_calibration_summary(results) -> None:
    """Print a concise, MATLAB-aligned calibration summary."""

    print("\nCalibration Summary")
    print("-------------------")
    print(f"Forward gyro model : {results.model['forward']['gyr']}")
    print(f"Inverse gyro model : {results.model['inverse']['gyr']}")
    print(f"Inverse acc model  : {results.model['inverse']['acc']}")
    print(f"Gg input term      : {results.model['gsens_term']}")

    bg = results.calib.get("bg")
    if bg is not None:
        print(f"bg [rad/s] : {_fmt_vec(bg)}")

    acc = results.calib.get("acc", {})
    if acc.get("ba") is not None:
        print(f"ba [m/s^2] : {_fmt_vec(acc['ba'])}")
    if acc.get("gravity_magnitude") is not None:
        print(f"gravity [m/s^2] : {float(acc['gravity_magnitude']):.6g}")
    if acc.get("Ca") is not None:
        print("Ca summary :")
        print(np.asarray(acc["Ca"]))
        print(f"rcond(Ca) : {_safe_rcond(acc['Ca']):.3e}")
    if acc.get("Sa") is not None:
        print("Sa summary :")
        print(np.asarray(acc["Sa"]))
    if acc.get("Ma") is not None:
        print("Ma summary :")
        print(np.asarray(acc["Ma"]))

    gyr = results.calib.get("gyr", {})
    if gyr.get("Cg") is not None:
        print("Cg summary :")
        print(np.asarray(gyr["Cg"]))
        print(f"rcond(Cg) : {_safe_rcond(gyr['Cg']):.3e}")
    if gyr.get("Kg") is not None:
        print("Kg summary :")
        print(np.asarray(gyr["Kg"]))
    if gyr.get("Mg") is not None:
        print("Mg summary :")
        print(np.asarray(gyr["Mg"]))
    if gyr.get("Gg") is not None:
        print("Gg summary :")
        print(np.asarray(gyr["Gg"]))

    summary = results.validation.get("summary", {}) if isinstance(results.validation, dict) else {}
    if summary:
        print("\nValidation summary:")
        _print_summary(summary)

    temp_block = results.calib.get("temp", {})
    if isinstance(temp_block, dict) and temp_block:
        _print_temperature_model_summary(temp_block)

    allan = results.analysis.get("allan") if isinstance(results.analysis, dict) else None
    if isinstance(allan, dict) and allan:
        print("\nAllan summary:")
        if isinstance(allan.get("gyro"), dict):
            _print_allan_block("gyro", allan["gyro"])
        if isinstance(allan.get("acc"), dict):
            _print_allan_block("acc", allan["acc"])

    print(f"\nAllan enabled      : {int(bool(results.options['pipeline'].get('enable_allan', False)))}")
    print(f"Temp model enabled : {int(bool(results.options['pipeline'].get('enable_temperature_model', False)))}")
    print(f"Auto segment       : {int(bool(results.options['pipeline'].get('enable_auto_segment', False)))}")
    if summary:
        print(f"Summary temp state : {get_temperature_model_status(temp_block)}")
        print(f"Summary Allan state: {get_allan_status(results.analysis)}")


def _print_summary(summary: dict) -> None:
    if "mode" in summary:
        print(f"  mode                            : {summary['mode']}")
    if "num_static_samples" in summary:
        print(f"  num_static_samples              : {summary['num_static_samples']}")
    if "available_blocks" in summary:
        print(f"  available_blocks                : {summary['available_blocks']}")
    if "completed_modules" in summary:
        print(f"  completed_modules               : {summary['completed_modules']}")
    if summary.get("static_gyro_mean_after_bias") is not None:
        print(f"  static_gyro_mean_after_bias     : {_fmt_vec(summary.get('static_gyro_mean_after_bias'))}")
    if summary.get("static_gyro_rms_after_bias") is not None:
        print(f"  static_gyro_rms_after_bias      : {_fmt_vec(summary.get('static_gyro_rms_after_bias'))}")
    if "static_acc_norm_mean" in summary and summary.get("static_acc_norm_mean") is not None:
        print(f"  static_acc_norm_mean            : {summary['static_acc_norm_mean']:.6g}")
    if "static_acc_norm_std" in summary and summary.get("static_acc_norm_std") is not None:
        print(f"  static_acc_norm_std             : {summary['static_acc_norm_std']:.6g}")
    if summary.get("gyro_std") is not None:
        print(f"  gyro_std [rad/s]                : {_fmt_vec(summary.get('gyro_std'))}")
    if summary.get("acc_std") is not None:
        print(f"  acc_std [m/s^2]                 : {_fmt_vec(summary.get('acc_std'))}")
    if "temperature_model_status" in summary:
        print(f"  temperature_model_status        : {summary['temperature_model_status']}")
    if "allan_status" in summary:
        print(f"  allan_status                    : {summary['allan_status']}")
    if "Ca_rcond" in summary and summary.get("Ca_rcond") is not None:
        print(f"  rcond(Ca)                       : {summary['Ca_rcond']:.3e}")
    if "Cg_rcond" in summary and summary.get("Cg_rcond") is not None:
        print(f"  rcond(Cg)                       : {summary['Cg_rcond']:.3e}")


def _print_allan_block(name: str, block: dict) -> None:
    estimate = block.get("estimate", {}) if isinstance(block.get("estimate"), dict) else {}
    print(f"  {name}:")
    print(f"    valid                        : {int(bool(block.get('valid', False)))}")
    print(f"    confidence                   : {estimate.get('confidence', block.get('confidence', 'n/a'))}")
    print(f"    num_samples                  : {block.get('num_samples', 0)}")
    tau = block.get("tau")
    if tau is not None and len(np.asarray(tau).reshape(-1)) > 0:
        tau = np.asarray(tau, dtype=float).reshape(-1)
        print(f"    tau range [s]                : {tau[0]:.6g} -> {tau[-1]:.6g}")
    if "noise_density" in estimate:
        print(f"    noise_density                : {_fmt_vec(estimate.get('noise_density'))}")
    if "bias_instability" in estimate:
        print(f"    bias_instability             : {_fmt_vec(estimate.get('bias_instability'))}")
    if "random_walk" in estimate:
        print(f"    random_walk                  : {_fmt_vec(estimate.get('random_walk'))}")
    if "message" in block:
        print(f"    message                      : {block.get('message', '')}")


def _print_temperature_model_summary(temp_block: dict) -> None:
    model_file = temp_block.get("model_file", temp_block.get("modelFile", {}))
    if not isinstance(model_file, dict):
        model_file = {}
    if model_file:
        print("\nTemperature model summary:")
        print(f"  reference_temperature   : {_fmt_scalar(model_file.get('reference_temperature'))}")
        print(f"  temperature_range       : {_fmt_range(model_file.get('temperature_range'))}")
        print(f"  extrapolation_mode      : {model_file.get('extrapolation_mode', '')}")
    for label, model in (
        ("gyro", temp_block.get("bg_model", temp_block.get("bgModel"))),
        ("acc", temp_block.get("ba_model", temp_block.get("baModel"))),
    ):
        if isinstance(model, dict) and model:
            print(f"  {label}:")
            print(f"    valid                        : {int(bool(model.get('valid', False)))}")
            print(f"    low_confidence               : {int(bool(model.get('low_confidence', False)))}")
            print(f"    method                       : {model.get('method', '')}")
            print(f"    type                         : {model.get('type', '')}")
            metrics = model.get("metrics", {}) if isinstance(model.get("metrics"), dict) else {}
            print(f"    temp_span                    : {_fmt_scalar(metrics.get('temp_span'))}")
            print(f"    num_points                   : {metrics.get('num_points', 0)}")
            print(f"    num_bins                     : {metrics.get('num_bins', 0)}")
            print(f"    rmse                         : {_fmt_scalar(metrics.get('rmse'))}")
            print(f"    max_abs_residual             : {_fmt_scalar(metrics.get('max_abs_residual'))}")
            if label == "gyro":
                print(f"    reference_bg [rad/s]         : {_fmt_vec(model.get('reference_bg'))}")
                print(f"    residual_rms [rad/s]         : {_fmt_vec(model.get('residual_rms'))}")
            else:
                print(f"    reference_ba [m/s^2]         : {_fmt_vec(model.get('reference_ba'))}")
                print(f"    residual_rms [m/s^2]         : {_fmt_vec(model.get('residual_rms'))}")
            print(f"    message                      : {model.get('message', '')}")


def _safe_rcond(matrix) -> float:
    arr = np.asarray(matrix, dtype=float)
    return float(1.0 / np.linalg.cond(arr))


def _fmt_vec(value) -> str:
    if value is None:
        return "[]"
    arr = np.asarray(value, dtype=float).reshape(-1)
    if arr.size == 0:
        return "[]"
    return "[" + ", ".join(f"{x:.6g}" for x in arr) + "]"


def _fmt_scalar(value) -> str:
    if value is None:
        return "n/a"
    try:
        scalar = float(value)
    except Exception:
        return str(value)
    if np.isnan(scalar):
        return "nan"
    return f"{scalar:.6g}"


def _fmt_range(value) -> str:
    if not isinstance(value, (list, tuple)) or len(value) != 2:
        return "[]"
    return "[" + ", ".join(_fmt_scalar(item) for item in value) + "]"

File: imu_calib_matlab/src/runtime/print_calibration_summary.m
function print_calibration_summary(results)
%PRINT_CALIBRATION_SUMMARY Print a concise summary of calibration results.

if nargin < 1 || ~isstruct(results)
    error('print_calibration_summary:InvalidInput', ...
        'results must be a struct.');
end

fprintf('\nCalibration Summary\n');
fprintf('-------------------\n');
summaryPayload = build_result_summary(results);

if isfield(results, 'model')
    fprintf('Forward gyro model : %s\n', results.model.forward.gyr);
    fprintf('Inverse gyro model : %s\n', results.model.inverse.gyr);
    fprintf('Forward acc model  : %s\n', results.model.forward.acc);
    fprintf('Inverse acc model  : %s\n', results.model.inverse.acc);
    fprintf('Gg input term      : %s\n', results.model.gsens_term);
end

if isfield(results, 'calib') && isfield(results.calib, 'bg') && ~isempty(results.calib.bg)
    print_vector3('bg [rad/s]', results.calib.bg);
end

if isfield(results, 'calib') && isfield(results.calib, 'acc')
    acc = results.calib.acc;
    if isfield(acc, 'Ca') && ~isempty(acc.Ca)
        print_matrix3('Ca', acc.Ca);
        fprintf('rcond(Ca) = %.3e\n', rcond(acc.Ca));
    end
    if isfield(acc, 'Sa') && ~isempty(acc.Sa)
        print_matrix3('Sa', acc.Sa);
    end
    if isfield(acc, 'Ma') && ~isempty(acc.Ma)
        print_matrix3('Ma', acc.Ma);
    end
    if isfield(acc, 'ba') && ~isempty(acc.ba)
        print_vector3('ba [m/s^2]', acc.ba);
    end
    if isfield(acc, 'gravity_magnitude') && ~isempty(acc.gravity_magnitude)
        fprintf('gravity_magnitude  : %.6g\n', acc.gravity_magnitude);
    end
end

if isfield(results, 'calib') && isfield(results.calib, 'gyr')
    gyr = results.calib.gyr;
    if isfield(gyr, 'Cg') && ~isempty(gyr.Cg)
        print_matrix3('Cg', gyr.Cg);
        fprintf('rcond(Cg) = %.3e\n', rcond(gyr.Cg));
    end
    if isfield(gyr, 'Kg') && ~isempty(gyr.Kg)
        print_matrix3('Kg', gyr.Kg);
    end
    if isfield(gyr, 'Mg') && ~isempty(gyr.Mg)
        print_matrix3('Mg', gyr.Mg);
    end
    if isfield(gyr, 'Gg') && ~isempty(gyr.Gg)
        print_matrix3('Gg', gyr.Gg);
    end
end

if isfield(results, 'validation') && isfield(results.validation, 'summary')
    fprintf('Validation summary:\n');
    print_expanded_validation_summary(results.validation.summary);
end

if isfield(results, 'calib') && isfield(results.calib, 'temp') && isstruct(results.calib.temp)
    print_temperature_model_summary(getfield_or_default(results.calib.temp, 'bgModel', struct()), 'bgModel');
    print_temperature_model_summary(getfield_or_default(results.calib.temp, 'baModel', struct()), 'baModel');
    if isfield(results.calib.temp, 'message') && ~isempty(results.calib.temp.message)
        fprintf('Temperature model message: %s\n', char(string(results.calib.temp.message)));
    end
end

if isfield(results, 'analysis') && isfield(results.analysis, 'allan') && isstruct(results.analysis.allan)
    print_allan_summary(results.analysis.allan);
end
if isstruct(summaryPayload) && isfield(summaryPayload, 'core_outputs')
    fprintf('Summary temp state : %s\n', char(string(summaryPayload.core_outputs.temperature_model_status)));
    fprintf('Summary Allan state: %s\n', char(string(summaryPayload.core_outputs.allan_status)));
end
end

function print_expanded_validation_summary(summary)
if isfield(summary, 'mode') && ~isempty(summary.mode)
    fprintf('  mode                        : %s\n', char(string(summary.mode)));
end
if isfield(summary, 'num_static_samples') && ~isempty(summary.num_static_samples)
    fprintf('  num_static_samples          : %d\n', summary.num_static_samples);
end
if isfield(summary, 'available_blocks') && ~isempty(summary.available_blocks)
    fprintf('  available_blocks            : %s\n', stringify_list(summary.available_blocks));
end
if isfield(summary, 'completed_modules') && ~isempty(summary.completed_modules)
    fprintf('  completed_modules           : %s\n', stringify_list(summary.completed_modules));
end
if isfield(summary, 'static_gyro_mean_after_bias') && ~isempty(summary.static_gyro_mean_after_bias)
    print_vector3('  static_gyro_mean_after_bias [rad/s]', summary.static_gyro_mean_after_bias);
end
if isfield(summary, 'static_gyro_rms_after_bias') && ~isempty(summary.static_gyro_rms_after_bias)
    print_vector3('  static_gyro_rms_after_bias [rad/s]', summary.static_gyro_rms_after_bias);
end
print_scalar_if_present(summary, 'static_acc_norm_mean', '  static_acc_norm_mean [m/s^2]');
print_scalar_if_present(summary, 'static_acc_norm_std', '  static_acc_norm_std [m/s^2]');
print_scalar_if_present(summary, 'static_acc_norm_mean_before', '  static_acc_norm_mean_before [m/s^2]');
print_scalar_if_present(summary, 'static_acc_norm_mean_after', '  static_acc_norm_mean_after [m/s^2]');
print_scalar_if_present(summary, 'static_acc_norm_std_before', '  static_acc_norm_std_before [m/s^2]');
print_scalar_if_present(summary, 'static_acc_norm_std_after', '  static_acc_norm_std_after [m/s^2]');
if isfield(summary, 'gyro_std') && ~isempty(summary.gyro_std)
    print_vector3('  gyro_std [rad/s]', summary.gyro_std);
end
if isfield(summary, 'acc_std') && ~isempty(summary.acc_std)
    print_vector3('  acc_std [m/s^2]', summary.acc_std);
end
if isfield(summary, 'temperature_model_status') && ~isempty(summary.temperature_model_status)
    fprintf('  temperature_model_status    : %s\n', char(string(summary.temperature_model_status)));
end
if isfield(summary, 'allan_status') && ~isempty(summary.allan_status)
    fprintf('  allan_status                : %s\n', char(string(summary.allan_status)));
end
end

function print_temperature_model_summary(model, label)
if ~isstruct(model) || isempty(fieldnames(model))
    return;
end

fprintf('%s:\n', label);
print_scalar_if_present(model, 'valid', '  valid');
print_scalar_if_present(model, 'low_confidence', '  low_confidence');
if isfield(model, 'method') && ~isempty(model.method)
    fprintf('  method                     : %s\n', char(string(model.method)));
end
if isfield(model, 'type') && ~isempty(model.type)
    fprintf('  type                       : %s\n', char(string(model.type)));
end
if isfield(model, 'reference_temperature') && ~isempty(model.reference_temperature) && isfinite(model.reference_temperature)
    fprintf('  reference_temperature      : %.6g\n', model.reference_temperature);
end
if isfield(model, 'temperature_range') && ~isempty(model.temperature_range)
    tempRange = double(model.temperature_range(:));
    if numel(tempRange) == 2 && all(isfinite(tempRange))
        fprintf('  temperature_range          : [%.6g, %.6g]\n', tempRange(1), tempRange(2));
    end
end
if isfield(model, 'extrapolation_mode') && ~isempty(model.extrapolation_mode)
    fprintf('  extrapolation_mode         : %s\n', char(string(model.extrapolation_mode)));
end
if isfield(model, 'metrics') && isstruct(model.metrics)
    metrics = model.metrics;
    print_scalar_if_present(metrics, 'temp_span', '  temp_span');
    print_scalar_if_present(metrics, 'num_points', '  num_points');
    print_scalar_if_present(metrics, 'num_bins', '  num_bins');
    print_scalar_if_present(metrics, 'rmse', '  rmse');
    print_scalar_if_present(metrics, 'max_abs_residual', '  max_abs_residual');
end
if isfield(model, 'reference_bg') && ~isempty(model.reference_bg)
    print_vector3('  reference_bg', model.reference_bg);
end
if isfield(model, 'reference_ba') && ~isempty(model.reference_ba)
    print_vector3('  reference_ba', model.reference_ba);
end
if isfield(model, 'residual_rms') && ~isempty(model.residual_rms)
    print_vector3('  residual_rms', model.residual_rms);
end
if isfield(model, 'residual_std') && ~isempty(model.residual_std)
    print_vector3('  residual_std', model.residual_std);
end
if isfield(model, 'message') && ~isempty(model.message)
    fprintf('  message                    : %s\n', char(string(model.message)));
end
end

function print_allan_summary(allanStruct)
fprintf('Allan summary:\n');
if isfield(allanStruct, 'gyro') && isstruct(allanStruct.gyro)
    print_single_allan_summary('gyro', allanStruct.gyro);
end
if isfield(allanStruct, 'acc') && isstruct(allanStruct.acc)
    print_single_allan_summary('acc', allanStruct.acc);
end
end

function print_single_allan_summary(name, allan)
fprintf('  %s:\n', name);
print_scalar_if_present(allan, 'valid', '    valid');
print_scalar_if_present(allan, 'num_samples', '    num_samples');
if isfield(allan, 'tau') && ~isempty(allan.tau)
    fprintf('    tau range [s]            : %.6g -> %.6g (%d points)\n', ...
        allan.tau(1), allan.tau(end), numel(allan.tau));
end
if isfield(allan, 'estimate') && isstruct(allan.estimate)
    est = allan.estimate;
    if isfield(est, 'noise_density') && ~isempty(est.noise_density)
        print_vector3('    noise_density', est.noise_density);
    end
    if isfield(est, 'bias_instability') && ~isempty(est.bias_instability)
        print_vector3('    bias_instability', est.bias_instability);
    end
    if isfield(est, 'random_walk') && ~isempty(est.random_walk)
        print_vector3('    random_walk', est.random_walk);
    end
    if isfield(est, 'confidence') && ~isempty(est.confidence)
        fprintf('    confidence               : %s\n', char(string(est.confidence)));
    end
end
if isfield(allan, 'message') && ~isempty(allan.message)
    fprintf('    message                  : %s\n', char(string(allan.message)));
end
end

function print_vector3(label, value)
v = double(value(:));
if numel(v) == 3
    fprintf('%s : [%.6g, %.6g, %.6g]\n', label, v(1), v(2), v(3));
else
    fprintf('%s : ', label);
    disp(v.');
end
end

function print_matrix3(label, M)
A = double(M);
fprintf('%s :\n', label);
disp(A);
end

function print_scalar_if_present(S, fieldName, label)
if isstruct(S) && isfield(S, fieldName) && ~isempty(S.(fieldName))
    value = S.(fieldName);
    if islogical(value)
        value = double(value);
    end
    if isscalar(value) && isnumeric(value) && isfinite(value)
        fprintf('%s : %.6g\n', label, double(value));
    end
end
end

function text = stringify_list(value)
if iscell(value)
    text = strjoin(cellfun(@(x) char(string(x)), value, 'UniformOutput', false), ', ');
else
    text = char(string(value));
end
end
